standard/patterns.md

```diff
@@ -0,0 +1,176 @@
+I would suggest placing this new chapter between the existing 12 (Statements) and 13 (namespaces)
+
+# §patterns-new-clause Patterns and pattern matching
+
+## §patterns-new-clause-general General
+
+A ***pattern*** is a syntactic form that can be used with the `is` operator ([§11.11.11](expressions.md#111111-the-is-operator)) and in a *switch_statement* ([§12.8.3](statements.md#1283-the-switch-statement)) to express the shape of data against which incoming data is to be compared. A pattern is tested against the *expression* of a switch statement, or against a *relational_expression* that is on the left-hand side of an `is` operator. We call this a ***pattern input value***.
+
+## §patterns-new-clause-forms Pattern Forms
+
+A pattern may have one of the following forms:
+
+```ANTLR
+pattern
+    : declaration_pattern
+    | constant_pattern
+    | var_pattern
+    ;
+```
+
+A *declaration_pattern* and a *var_pattern* can result in the declaration of a local variable.
+
+Each pattern form defines the set of types for input values that the pattern may be applied to.  We say a pattern `P` is *applicable to* a type `T` if `T` is among the types whose values the pattern may match.  It is an error if a pattern `P` appears in a program to match a *pattern input value* of type `T` if `P` is not applicable to `T`.
+
+Each pattern form defines the set of values for which the pattern *matches* the value.
+
+### §declaration-pattern-new-clause Declaration pattern
+
+A *declaration_pattern* is used to test that a value has a given type and, if the test succeeds, provide the value in a variable of that type.
```

Comment:
From below its "a value has a type which is pattern compatible with a given type" – which is a bit long winded so some wordsmithing is required…

Reply:
This would be better done sooner rather than later.  Two things:
1. The specification about the value being of the given type is actually a bit more complex and should delegate to the spec of the is-type operator.  This is a runtime type test.
2. The *pattern compatible* concept is to give warnings based on compile-time types that are impossible to match.  If that isn't clear, it should be made clear.

Reply:
Addressed in #934

Added text that the runtime test must produced the same results as the is-type operator. Earlier edits should have already clarified the *pattern compatible* concept as a compile-time operation.